### PR TITLE
DRILL-5795: Parquet Filter push down at rowgroup level

### PR DESCRIPTION
**Pull request by dprofeta**
Before this commit, the filter was pruning complete files. When a file
is composed of multiple rowgroups, it was not able to prune one
rowgroup from the file. Now, when the filter find that a rowgroup
doesn't match it will be remove from the scan.

### CODE REVIEW

**Comment by dprofeta** [comment]
@parthchandra Can you please review?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java [hunk 1]
```diff
@@ -819,63 +827,64 @@ private void init() throws IOException {
...
-          Long previousCount = columnValueCounts.get(schemaPath);
-          if (previousCount != null) {
-            if (previousCount != GroupScan.NO_COLUMN_STATS) {
-              if (column.getNulls() != null) {
-                Long newCount = rowCount - column.getNulls();
-                columnValueCounts.put(schemaPath, columnValueCounts.get(schemaPath) + newCount);
-              }
-            }
-          } else {
+    for (RowGroupInfo rowGroup : this.rowGroupInfos) {
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java]: Isn't this doing the same thing as the original code? RowGroupInfos is built from the RowGroupMetadata in the files?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java [hunk 2]
```diff
@@ -1095,7 +1104,7 @@ public GroupScan applyFilter(LogicalExpression filterExpr, UdfUtilities udfUtili
 
     final Set<SchemaPath> schemaPathsInExpr = filterExpr.accept(new ParquetRGFilterEvaluator.FieldReferenceFinder(), null);
 
-    final List<RowGroupMetadata> qualifiedRGs = new ArrayList<>(parquetTableMetadata.getFiles().size());
+    final List<RowGroupInfo> qualifiedRGs = new ArrayList<>(rowGroupInfos.size());
```
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java]: Never mind the previous comment. It's probably better to use RowGroupInfos throughout the code.

**Comment by dprofeta** [comment]
I will add a unit test to test the number of rowgroups that are scanned by the groupscan to see if the filter is well able to prune rowgroup.

**Comment by dprofeta** [comment]
I have updated the patch with a unit test and fixed an issue when everything is filtered.

**Comment by paul-rogers** [comment]
Please resolve commits and rebase onto master.

**Comment by dprofeta** [comment]
@paul-rogers done

**Comment by paul-rogers** [comment]
This change causes one of our functional tests to fail. We will have to track down the issue and either update the test, or post the problem here.

**Comment by priteshm** [comment]
@paul-rogers, @kkhatua can you provide some more information on the test case that failed? Hopefully, @dprofeta can replicate it in his environment.

**Comment by cchang738** [comment]
There is a plan verification failure due to plan change. The plan baseline needs to be changed after this PR is merged.

Plan Verification Failures:
/root/drillAutomation/mapr/framework/resources/Functional/int96/q28.q
Query: 
explain plan for select voter_id, name from `hive1_parquet_part` where date_part('year', create_timestamp1)=2018

Expected and actual text plans are different.
Expected:
.*numFiles=2, usedMetadataFile=true.*

Actual:
00-00    Screen
00-01      Project(voter_id=[$0], name=[$1])
00-02        Project(voter_id=[$1], name=[$2])
00-03          Scan(groupscan=[ParquetGroupScan [entries=[ReadEntryWithPath [path=/drill/testdata/subqueries/hive1_parquet_part/0_0_10.parquet], ReadEntryWithPath [path=/drill/testdata/subqueries/hive1_parquet_part/0_0_9.parquet]], selectionRoot=/drill/testdata/subqueries/hive1_parquet_part, numFiles=2, numRowGroups=2, usedMetadataFile=true, cacheFileRoot=/drill/testdata/subqueries/hive1_parquet_part, columns=[`create_timestamp1`, `voter_id`, `name`]]])

**Comment by dprofeta** [comment]
Indeed, in the plan, there is an added information that is the number of rowgroup scanned. I think the check of the plan should be fixed by checking separately numFiles and usedMetadataFile without expecting they are next to each other.

**Pull request closed without merging**